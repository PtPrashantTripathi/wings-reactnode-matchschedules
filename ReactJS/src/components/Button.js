// Button component
// props:
//   onClick - a function calls native onClick event
//   children - renders as child
//   className - helps to style the component
const Button = ({ onClick, children, className }) => (
  <button onClick={onClick} className={className}>
    {children}
  </button>
);

export default Button;