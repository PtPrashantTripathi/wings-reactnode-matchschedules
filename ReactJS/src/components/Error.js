// Component that renders error message.
// props:
//   message: error message that has to be displayed
const Error = ({ message }) => (
  message ? <div className="error">{message}</div> : null
);


export default Error;