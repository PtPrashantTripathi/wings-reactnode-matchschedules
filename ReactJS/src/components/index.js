export { default as List } from './List';
export { default as Card } from './Card';
export { default as Modal } from './Modal';
export { default as Button } from './Button';
export { default as Error } from './Error';
export { default as Form } from './Form';