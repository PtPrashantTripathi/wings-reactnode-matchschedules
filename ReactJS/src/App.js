import { Component } from "react";
import { name } from "../package.json";
import { List, Modal, Button, Error, Form } from "./components";

// Assign the backend base url here
export const NODE_APP_URL = "http://localhost:8001/fixtures";

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      data: [],
      error: undefined,
      showModal: false,
      formValues: {},
      formError: undefined
    }
  }

  componentDidMount = () => {
    this.fetchData();
  };

  // Function to Show the Modal component
  showModalHandler = () => {
     // Your code goes here
  }

  // Function to Hide the Modal component
  closeModalHandler = () => {
     // Your code goes here
  }

  // Handles all input entered in the form component
  // and stores the values in the state variable "formValues" of Object type
  // formValues is a key-value pair of input elements { name: "value", name1: "value1" }
  inputChangeHandler = (e) => {
    const { name, value, type } = e.target;
    let { formValues } = this.state;
    formValues = { ...formValues, [name]: type === "number" ? parseInt(value) : value };
    this.setState({ formValues });
  }

  // Fetch data from the api
  // use NODE_APP_URL variable for the url in fetch method
  // NODE_APP_URL variable is assigned in the top of this file (src/App.js).
  fetchData = async () => {
    // Your code goes here
    // Fill up the code required for posting data to backend
  };

  // SubmitHandler should be used to create a record i.e., to execute post request to backend
  // On success of post request close modal and fetch call fetchData method again.
  // On Error set the error message in the banner.
  submitHandler = (e) => {
    e.preventDefault();
    const { formValues } = this.state;

    fetch(NODE_APP_URL, {
      // Your code goes here
      // Fill up the params required for posting data to backend
    })
      .then((response) => response.json())
      .then((res) => {
        if (res.error) this.setState({ formError: res.error });
        else {
          this.fetchData();
          this.closeModalHandler();
        }
      });
  }

  render() {
    const { showModal, error, data, formError } = this.state;

    return (
      <div className="app">
        <div className="app-body">
          <h2 className="app-title">{name.replace(/_/g, ' ')}</h2>
          <Error message={error} />
          <List data={data} />

          <div className="footer-controls">
            {/* Your code goes here */}
            {/* Render a Button that will display the Modal */}
          </div>
        </div>

        <Modal show={showModal} closeHandler={this.closeModalHandler}>
          {/* Your code goes here */}
          {/* Render the Form component here */}
        </Modal>
      </div>
    );
  }
}

export default App;
