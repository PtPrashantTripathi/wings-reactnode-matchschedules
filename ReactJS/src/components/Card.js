import Button from "./Button";

// Card component renders each record in "data.records" from the api
// consists of table, thead, tbody elements along with tr, th, td
// props:
//   deleteHandler: a function that deletes the record
//   _id: _id from the api 
//   index: each row index 
//   count: count from the api
//   venue: venue key from the api
//   team1: team1 key from the api
//   team2: team2 key from the api
//   date: date key from the api. Make sure to render it in format "Day Mon DD YYYY" ex: "Sun Jun 23 2021"

const Card = ({
  _id,
  index,
  count,
  venue,
  team1,
  team2,
  date
}) => (
    <div className="card" id={_id}>
      <div className="card-header">
        <span>
          <span className="small-txt">Match </span>
          {index}
          <span className="small-txt"> of </span>
          {count}
        </span>
        <span>
          <span className="small-txt">Venue: </span>
          {venue}
        </span>
      </div>
      <span id="teams">
        {team1}
        <span className="small-txt"> vs </span>
        {team2}
      </span>
      <div className="card-footer">
        <span id="matchDate" className="small-txt">
          {/* Render the date here in format "Day Mon DD YYYY" (ex: "Sun Jun 23 2021") */}
          {new Date(date).toDateString()}
        </span>
      </div>
    </div>
  );

export default Card;