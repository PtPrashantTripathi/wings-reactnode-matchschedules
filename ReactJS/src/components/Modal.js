import Button from "./Button";

// Renders a Modal
// props:
//   show: prop willÂ can be used to show/hide the modal
//   closeHandler: prop is used by Button, will close the modal
const Modal = ({ children, show, closeHandler }) => (
  <>
    {show ?
      <div className="modal-bg">
        <div className="modal">
          <Button onClick={closeHandler} className="btn-close">&#x2715;</Button>
          {children}
        </div>
      </div> : null
    }
  </>
);

export default Modal;