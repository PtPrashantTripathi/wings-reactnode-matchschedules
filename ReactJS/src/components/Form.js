import model from '../model';
import Button from "./Button";
import Error from "./Error";

// Form Component has following elements
// use label and input elements for each input field
// custom Button component executes "onSaveHandler" prop when clicked
// props:
//   title: renders heading using h3 tag
//   onSaveHandler: called when the Save button is clicked
//   inputOnChangeHandler: helps to capture the entered data to store it in state
//   error: renders the error from backend
const Form = ({ title, inputOnChangeHandler, onSaveHandler, error }) => (
  <form>
    <h3>{title}</h3>
    <Error message={error} />
    {model.filter(i => i.input_type)
      .map(i =>
        <div key={i.key} className="form-group">
          <label htmlFor={i.column_name}>{i.title}</label>
          <input
            type={i.input_type}
            name={i.column_name}
          />
        </div>
      )}
    <Button onClick={onSaveHandler} className="success">Save</Button>
  </form>
);

export default Form;