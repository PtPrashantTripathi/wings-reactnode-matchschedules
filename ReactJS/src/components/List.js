import Card from "./Card";


// props:
//   data: consists of data records to be rendered
const List = ({ data }) => (
  data ? (<div className="list">
    {/* Your code goes here */}
    {/* Render the Card with required props here */}
  </div>) : null
);


export default List;