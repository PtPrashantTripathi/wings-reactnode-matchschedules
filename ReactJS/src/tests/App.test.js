import { create } from 'react-test-renderer';
import App, { NODE_APP_URL } from '../App';
import testData from './testData';
import { List, Modal, Button, Error, Form } from '../components';
import { name } from '../../package.json';

describe('App', () => {
  beforeEach(() => {
    const mockSuccessResponse = testData;
    const mockJsonPromise = Promise.resolve(mockSuccessResponse);
    const mockFetchPromise = Promise.resolve({
      json: () => mockJsonPromise,
    });
    jest.spyOn(global, "fetch").mockImplementation(() => mockFetchPromise);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('renders app component', () => {
    const component = create(<App />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  test("component did mount", (done) => {
    create(<App />);

    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(global.fetch).toHaveBeenCalledWith(
      NODE_APP_URL
    );
    global.fetch.mockClear();
    done();
  });

  test('renders App correctly', () => {
    const component = create(<App />).root;

    expect(component.findByType(List)).not.toBeNull();
    expect(component.findByType(Modal)).not.toBeNull();
    expect(component.findByType(Button)).not.toBeNull();
    expect(component.findByType(Error)).not.toBeNull();
  });

  test("App Button click opens Modal and renders form", () => {
    const component = create(<App />);
    const instance = component.getInstance();
    const button = component.root.findByType(Button);

    expect(instance.state.showModal).toBe(false);

    expect(button.props.onClick).not.toBe(undefined);
    expect(button.props.onClick).toEqual(instance.showModalHandler);
    button.props.onClick();

    expect(instance.state.showModal).toBe(true);

    // Expect Modal and Form to have rendered by now
    const modalComponent = component.root.findByType(Modal);
    expect(modalComponent).not.toBeNull();
    expect(modalComponent.props.closeHandler).toEqual(instance.closeModalHandler);
    expect(modalComponent.props.show).toEqual(instance.state.showModal);

    const formComponent = component.root.findByType(Form);
    expect(formComponent).not.toBeNull();
    expect(formComponent.props.onSaveHandler).toEqual(instance.submitHandler);
    expect(formComponent.props.inputOnChangeHandler).toEqual(instance.inputChangeHandler);
  });

  test("App showModalHandler", () => {
    const component = create(<App />);
    const instance = component.getInstance();

    expect(instance.state.showModal).toBe(false);

    instance.showModalHandler();
    expect(instance.state.showModal).toBe(true);
    expect(component.root.findByType(Modal)).not.toBeNull();
    expect(component.root.findByType(Form)).not.toBeNull();
  });

  test("App closeModalHandler", () => {
    const component = create(<App />);
    const instance = component.getInstance();

    instance.setState({
      showModal: true,
      error: "Error",
      formError: "Error",
      formValues: { "abc": "def" }
    });

    instance.closeModalHandler();
    expect(instance.state.showModal).toBe(false);
    expect(instance.state.formValues).toEqual({});
    expect(instance.state.formError).toEqual(undefined);
    expect(instance.state.error).toEqual(undefined);
  });

  test("App fetchData", () => {
    const component = create(<App />);
    const instance = component.getInstance();

    instance.fetchData();
    expect(global.fetch).toHaveBeenCalledTimes(2);
    expect(global.fetch).toHaveBeenCalledWith(NODE_APP_URL);
    global.fetch.mockClear();

    expect(instance.state.error).toEqual(undefined);
  });

  test("App fetchData return error", () => {
    const component = create(<App />);
    const instance = component.getInstance();

    const mockErrorResponse = "Some Error";
    const mockJsonErrorPromise = Promise.resolve({ error: mockErrorResponse });
    const mockFetchPromiseError = Promise.resolve({
      json: () => Promise.resolve(mockJsonErrorPromise),
    });
    jest.spyOn(global, "fetch").mockReturnValue(mockFetchPromiseError);

    instance.fetchData();
    expect(global.fetch).toHaveBeenCalledTimes(2);
    expect(global.fetch).toHaveBeenCalledWith(NODE_APP_URL);
    global.fetch.mockClear();

    expect(component.root.findByType(Error)).not.toBeNull();
  });

  test("App submitHandler", () => {
    const event = { target: { name: "abc", value: "123", type: 'number' }, preventDefault: jest.fn() };
    const formValues = { abc: 123 };
    const component = create(<App />);
    const instance = component.getInstance();

    instance.inputChangeHandler(event);
    instance.submitHandler(event);

    expect(global.fetch).toHaveBeenCalledTimes(2);
    expect(global.fetch).toHaveBeenCalledWith(NODE_APP_URL);
    expect(global.fetch).toHaveBeenCalledWith(
      NODE_APP_URL,
      {
        method: "post",
        body: JSON.stringify(formValues),
        headers: { "Content-Type": "application/json" },
      }
    );
    global.fetch.mockClear();
  });

  test("App submitHandler returns error", () => {
    const event = { target: { name: "abc", value: "123", type: 'number' }, preventDefault: jest.fn() };
    const formValues = { abc: 123 };
    const component = create(<App />);
    const instance = component.getInstance();

    const mockErrorResponse = "Some Error";
    const mockJsonErrorPromise = Promise.resolve({ error: mockErrorResponse });
    const mockFetchPromiseError = Promise.resolve({
      json: () => Promise.resolve(mockJsonErrorPromise),
    });
    jest.spyOn(global, "fetch").mockReturnValue(mockFetchPromiseError);

    instance.inputChangeHandler(event);
    instance.submitHandler(event);

    expect(global.fetch).toHaveBeenCalledTimes(2);
    expect(global.fetch).toHaveBeenCalledWith(NODE_APP_URL);
    expect(global.fetch).toHaveBeenCalledWith(
      NODE_APP_URL,
      {
        method: "post",
        body: JSON.stringify(formValues),
        headers: { "Content-Type": "application/json" },
      }
    );
    global.fetch.mockClear();

    const errorComponent = component.root.findByType(Error);
    expect(errorComponent).not.toBeNull();
  });
});

