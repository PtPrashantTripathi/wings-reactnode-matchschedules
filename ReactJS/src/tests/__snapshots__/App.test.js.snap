// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders app component 1`] = `
<div
  className="app"
>
  <div
    className="app-body"
  >
    <h2
      className="app-title"
    >
      match schedule
    </h2>
    <div
      className="list"
    />
    <div
      className="footer-controls"
    >
      <button
        className="success"
        onClick={[Function]}
      >
        Add
      </button>
    </div>
  </div>
</div>
`;
